sm-gen/src/main/kotlin/org.smartregister.fhir.structuremaptool/Utils.kt

```diff
@@ -104,73 +114,67 @@ class Group(
             // TODO: Fix these to use infer
-            if (fieldPath.equals("id")) {
-                return "create('id') as id, id.value = evaluate(src, ${"$"}this.item${getPropertyPath()}.where(linkId = '$responseFieldId').answer.value)";
-            } else if (fieldPath.equals("rank")) {
-                return "create('positiveInt') as rank, rank.value = evaluate(src, ${"$"}this.item${getPropertyPath()}.where(linkId = '$responseFieldId').answer.value)";
+            if (fieldPath == "id" || fieldPath == "rank") {
+                expression = "create('${if (fieldPath == "id") "id" else "positiveInt"}') as $fieldPath, $fieldPath.value = evaluate(src, $expression)"
             } else {
 
                 // TODO: Infer the resource property type and answer to perform other conversions
                 // TODO: Extend this to cover other corner cases
-                var expression = "${"$"}this.item${getPropertyPath()}.where(linkId = '$responseFieldId').answer.value"
-
                 if (expression.isCoding(questionnaireResponse) && fieldPath.isEnumeration(this)) {
                     expression = expression.replace("answer.value", "answer.value.code")
                 } else if (inferType(fullPropertyPath()) == "CodeableConcept") {
-                    return "''";
+                    return "''"
                 }
-
-                return "evaluate(src, $expression)"
-
+                expression = "evaluate(src, $expression)"
             }
+            return expression
         }
 
         // 3. If it's a FHIR Path/StructureMap function, add the contents directly from here to the StructureMap
         if (fhirPathStructureMapFunctions != null && fhirPathStructureMapFunctions!!.isNotEmpty()) {
             // TODO: Fix the 2nd param inside the evaluate expression --> Not sure what this is but check this
             return fhirPathStructureMapFunctions!!
         }
-
-        // If it's a conversion
-        // 4. If the answer is a conversion, (Assume this means it's being convered to a reference)
+        // 4. If the answer is a conversion, (Assume this means it's being converted to a reference)
         if (conversion != null && conversion!!.isNotBlank() && conversion!!.isNotEmpty()) {
+            println("current resource to reference is $conversion")
+
             val resourceName = conversion!!.replace("$", "")
             var resourceIndex = conversion!!.replace("$$resourceName", "")
-
             if (resourceIndex.isNotEmpty()) {
                 resourceIndex = "[$resourceIndex]"
             }
-
-            // TODO: Create a GROUP that generates a reference to encapsulate this
-            //return "reference(evaluate(bundle, ${"$"}this.entry.where(resourceType = '$resourceName')$resourceIndex))"
-            return "reference(src)"
+            val reference = generateReference(resourceName = resourceName, resourceIndex = resourceIndex)
+            return reference
+            //return "reference(ref)"
```

Comment:
Do we want to delete this?

Reply:
the commented return is for the other version of referencing but I have it on a separate branch let me clean it away from here

Reply:
Cool. Also on the next PR you can jus return the `generateReference` instead of creating a val